File: packages/php/src/Services/STKPushService.php
<?php

namespace Yourdudeken\Mpesa\Services;

use Yourdudeken\Mpesa\Http\Client;
use Yourdudeken\Mpesa\Helpers\Signature;

class STKPushService
{
    private $client;
    private $signature;

    public function __construct(Client $client, Signature $signature)
    {
        $this->client = $client;
        $this->signature = $signature;
    }

    public function push($phonenumber, $amount, $account_number, $callbackurl = null, $transactionType = 'CustomerPayBillOnline', $shortCodeType = 'C2B')
    {
        $url = $this->client->getBaseUrl() . '/mpesa/stkpush/v1/processrequest';
        
        $data = [
            'BusinessShortCode' => config('mpesa.shortcode'),
            'Password' => $this->signature->lipaNaMpesaPassword(),
            'Timestamp' => $this->signature->getFormattedTimestamp(),
            'Amount' => (int) $amount,
            'PartyA' => $this->signature->phoneValidator($phonenumber),
            'PartyB' => $transactionType == 'CustomerPayBillOnline' 
                ? config('mpesa.shortcode') 
                : config('mpesa.till_number'),
            'TransactionType' => $transactionType,
            'PhoneNumber' => $this->signature->phoneValidator($phonenumber),
            'TransactionDesc' => 'Payment',
            'AccountReference' => $account_number,
            'CallBackURL' => $this->resolveCallbackUrl($callbackurl, 'callback_url'),
        ];

        return $this->client->post($url, $data, $shortCodeType);
    }

    public function query($checkoutRequestId, $shortCodeType = 'C2B')
    {
        $url = $this->client->getBaseUrl() . '/mpesa/stkpushquery/v1/query';
        
        $data = [
            'BusinessShortCode' => config('mpesa.shortcode'),
            'Password' => $this->signature->lipaNaMpesaPassword(),
            'Timestamp' => $this->signature->getFormattedTimestamp(),
            'CheckoutRequestID' => $checkoutRequestId,
        ];

        return $this->client->post($url, $data, $shortCodeType);
    }

    private function resolveCallbackUrl($paramUrl, $configKey)
    {
        $configUrl = config("mpesa.callbacks.{$configKey}");
        
        if ($paramUrl !== null) {
            return $paramUrl;
        } elseif ($configUrl !== null) {
            return $configUrl;
        }
        
        throw new \Exception("Ensure you have set the {$configKey} in the config or passed as a parameter");
    }
}

File: packages/go/mpesa/stk_push_service.go
package mpesa

import "fmt"

type STKPushService struct {
	httpClient *HTTPClient
	auth       *Auth
	helpers    *Helpers
	baseURL    string
}

func NewSTKPushService(httpClient *HTTPClient, auth *Auth, config *Config) *STKPushService {
	return &STKPushService{
		httpClient: httpClient,
		auth:       auth,
		helpers:    NewHelpers(config),
		baseURL:    httpClient.GetBaseURL(),
	}
}

func (s *STKPushService) Push(phonenumber string, amount int, accountNumber string, callbackURL string, transactionType string, shortCodeType string) (map[string]interface{}, error) {
	if accountNumber == "" {
		return nil, MissingAccountReference()
	}

	url := s.baseURL + "/mpesa/stkpush/v1/processrequest"
	data := map[string]interface{}{
		"BusinessShortCode": s.helpers.GetConfig("shortcode"),
		"Password":          s.helpers.LipaNaMpesaPassword(),
		"Timestamp":         s.helpers.GetFormattedTimestamp(),
		"Amount":            amount,
		"PartyA":            s.helpers.PhoneValidator(phonenumber),
		"PartyB":            s.helpers.GetConfig("shortcode"),
		"TransactionType":   transactionType,
		"PhoneNumber":       s.helpers.PhoneValidator(phonenumber),
		"TransactionDesc":   "Payment",
		"AccountReference":  accountNumber,
		"CallBackURL":       callbackURL,
	}

	token, err := s.auth.GetAccessToken(shortCodeType)
	if err != nil {
		return nil, err
	}

	return s.httpClient.Post(url, data, token)
}

func (s *STKPushService) Query(checkoutRequestID string, shortCodeType string) (map[string]interface{}, error) {
	url := s.baseURL + "/mpesa/stkpushquery/v1/query"
	data := map[string]interface{}{
		"BusinessShortCode": s.helpers.GetConfig("shortcode"),
		"Password":          s.helpers.LipaNaMpesaPassword(),
		"Timestamp":         s.helpers.GetFormattedTimestamp(),
		"CheckoutRequestID": checkoutRequestID,
	}

	token, err := s.auth.GetAccessToken(shortCodeType)
	if err != nil {
		return nil, err
	}

	return s.httpClient.Post(url, data, token)
}

func (s *STKPushService) Validate(accountNumber string) error {
	if accountNumber == "" {
		return fmt.Errorf("An Account Reference is required for All transactions.")
	}
	return nil
}
